File: SystemExtensions.FUnitTests/DatabaseTests.fs
module DatabaseTests

open System
open System.Data
open System.Threading.Tasks
open Psns.Common.Functional
open Foq
open NUnit.Framework
open FsUnit

type ext = Psns.Common.Functional.Prelude
type db = Psns.Common.SystemExtensions.Database.Prelude

// Connect Tests
let factory connection = Func<IDbConnection>(fun () -> connection)
let funFactory map = Func<IDbConnection, Try<string>> map

let ok _ = ext.Try(fun () -> "result")
let fail _ = ext.Try(fun () -> failwith "fail")

let eval map conn = db.Connect().Invoke(funFactory map, factory conn).Match((fun s -> s), (fun e -> e.Message))

let evalOk = eval ok
let evalFail = eval fail

[<Test>]
let ``it should return the result of the given function.`` () =
    let connection = Mock<IDbConnection>().Create()

    evalOk connection |> should equal "result"
    verify <@ connection.Dispose() @> once

[<Test>]
let ``it should return error text if action fails.`` () =
    let connection = Mock<IDbConnection>().Create()

    evalFail connection |> should equal "fail"
    verify <@ connection.Dispose() @> once

// ConnectAsync Tests
let asTask func = func >> Task.FromResult
let openAsync func = Func<IDbConnection, Task<IDbConnection>> func
let openOk = openAsync (asTask (fun cn -> cn))
let openFail = openAsync (fun _ -> failwith "fail")

let connectAsync conn openAsync map =
    db.ConnectAsync().Invoke(Func<IDbConnection, TryAsync<string>> map, openAsync, factory conn)

let evalAsync openAsync map conn = (connectAsync conn openAsync map).Match((fun s -> s), (fun e -> e.Message)).Result

let okAsync _ = ext.TryAsync(fun () -> Task.FromResult "result")
let failAsync _ = ext.TryAsync(fun () -> Task.FromResult (failwith "fail"))
let evalOkAsync = evalAsync openOk okAsync
let evalFailAsync = evalAsync openOk failAsync
let openAsyncFail = evalAsync openFail okAsync

[<Test>]
let ``it should return the result of the given async function.`` () =
    let connection = Mock<IDbConnection>().Create()

    evalOkAsync connection |> should equal "result"
    verify <@ connection.Dispose() @> once

[<Test>]
let ``it should return error text if openAsync fails.`` () =
    let connection = Mock<IDbConnection>().Create()

    openAsyncFail connection |> should equal "fail"
    verify <@ connection.Dispose() @> once

[<Test>]
let ``it should return error text if action async fails.`` () =
    let connection = Mock<IDbConnection>().Create()

    evalFailAsync connection |> should equal "fail"
    verify <@ connection.Dispose() @> once

File: SystemExtensions.UnitTests/DatabaseTests.cs
using Moq;
using NUnit.Framework;
using Psns.Common.Functional;
using System;
using System.Collections.Generic;
using System.Data;
using System.Threading.Tasks;
using static NUnit.StaticExpect.Expectations;
using static Psns.Common.Functional.Prelude;
using static Psns.Common.SystemExtensions.Database.Prelude;

namespace SystemExtensions.UnitTests.Database
{
    [TestFixture]
    public class DatabaseTests
    {
        [Test]
        public void Composing_Funcs_ReturnsResult()
        {
            var mockConnection = new Mock<IDbConnection>();
            var mockTransaction = new Mock<IDbTransaction>();
            var mockCommand = new Mock<IDbCommand>();
            var mockConnectionFactory = new Mock<Func<IDbConnection>>();
            mockConnection.Setup(c => c.BeginTransaction()).Returns(mockTransaction.Object);
            mockConnection.Setup(c => c.CreateCommand()).Returns(mockCommand.Object);
            mockTransaction.Setup(t => t.Connection).Returns(mockConnection.Object);
            mockConnectionFactory.Setup(f => f()).Returns(mockConnection.Object);

            var withCommand = fun((IDbCommand cmd) => "result");
            var createCommand = CreateCommand<string>().Par(withCommand);
            var beginTransaction = BeginTransaction<string>().Par(createCommand);
            var connect = Connect<string>().Par(beginTransaction).Compose(() => mockConnectionFactory.Object);

            Expect(connect().Match(r => r, e => e.Message), EqualTo("result"));

            mockConnection.Verify(c => c.Dispose());
            mockConnection.Verify(c => c.BeginTransaction());
            mockConnection.Verify(c => c.CreateCommand());
            mockTransaction.Verify(c => c.Dispose());
            mockTransaction.Verify(t => t.Connection);
            mockCommand.VerifySet(c => c.Transaction = mockTransaction.Object);
            mockCommand.Verify(c => c.Dispose());
        }

        [Test]
        public async Task Composing_AsyncFuncs_ReturnsResult()
        {
            var executionOrder = new Dictionary<string, int>();
            var executionCount = 0;
            var updateCount = act((string caller) => executionOrder[caller] = executionCount++);

            var mockConnection = new Mock<IDbConnection>();
            var mockTransaction = new Mock<IDbTransaction>();
            var mockCommand = new Mock<IDbCommand>();
            var mockConnectionFactory = new Mock<Func<IDbConnection>>();
            mockConnection.Setup(c => c.BeginTransaction())
                .Returns(mockTransaction.Object)
                .Callback(() => updateCount("trans"));
            mockConnection.Setup(c => c.CreateCommand())
                .Returns(mockCommand.Object)
                .Callback(() => updateCount("command"));
            mockConnection.Setup(c => c.Dispose()).Callback(() => updateCount("conDispose"));
            mockTransaction.Setup(t => t.Connection).Returns(mockConnection.Object);
            mockTransaction.Setup(t => t.Dispose()).Callback(() => updateCount("transDispose"));
            mockConnectionFactory.Setup(f => f()).Returns(mockConnection.Object);
            mockCommand.Setup(c => c.Dispose()).Callback(() => updateCount("comDispose"));

            var withCommand = fun((IDbCommand cmd) => { updateCount("with cmd"); return "result".AsTask(); });
            var createCommand = CreateCommandAsync<string>().Par(withCommand);
            var beginTransaction = BeginTransactionAsync<string>().Par(createCommand);
            var connect = ConnectAsync<string>()
                .Par(beginTransaction, conn => {updateCount("conn factory"); return conn.AsTask(); })
                .Compose(() => mockConnectionFactory.Object);

            Expect(await connect().Match(r => r, e => e.Message), EqualTo("result"));

            mockConnection.Verify(c => c.Dispose());
            mockConnection.Verify(c => c.BeginTransaction());
            mockConnection.Verify(c => c.CreateCommand());
            mockTransaction.Verify(c => c.Dispose());
            mockTransaction.Verify(t => t.Connection);
            mockCommand.VerifySet(c => c.Transaction = mockTransaction.Object);
            mockCommand.Verify(c => c.Dispose());

            Expect(executionOrder["conn factory"], EqualTo(0));
            Expect(executionOrder["trans"], EqualTo(1));
            Expect(executionOrder["command"], EqualTo(2));
            Expect(executionOrder["with cmd"], EqualTo(3));
            Expect(executionOrder["comDispose"], EqualTo(4));
            Expect(executionOrder["transDispose"], EqualTo(5));
            Expect(executionOrder["conDispose"], EqualTo(6));
        }

        [Test]
        public void Composing_Funcs_ExecutesInCorrectOrder()
        {
            var executionOrder = new Dictionary<string, int>();
            var executionCount = 0;
            var updateCount = act((string caller) => executionOrder[caller] = executionCount++);

            var mockConnection = new Mock<IDbConnection>();
            var mockTransaction = new Mock<IDbTransaction>();
            var mockCommand = new Mock<IDbCommand>();
            var mockConnectionFactory = new Mock<Func<IDbConnection>>();
            mockConnection.Setup(c => c.BeginTransaction())
                .Returns(mockTransaction.Object)
                .Callback(() => updateCount("trans"));
            mockConnection.Setup(c => c.CreateCommand())
                .Returns(mockCommand.Object)
                .Callback(() => updateCount("command"));
            mockConnection.Setup(c => c.Dispose()).Callback(() => updateCount("conDispose"));
            mockTransaction.Setup(t => t.Connection).Returns(mockConnection.Object);
            mockTransaction.Setup(t => t.Dispose()).Callback(() => updateCount("transDispose"));
            mockConnectionFactory.Setup(f => f()).Returns(mockConnection.Object);
            mockCommand.Setup(c => c.Dispose()).Callback(() => updateCount("comDispose"));

            var withCommand = fun((IDbCommand cmd) => { updateCount("with cmd"); return "result"; });
            var createCommand = CreateCommand<string>().Par(withCommand);
            var beginTransaction = BeginTransaction<string>().Par(createCommand);
            var openConnection = fun((Func<IDbConnection, Try<string>> func, IDbConnection conn) =>
                Try(() => conn.Tap(_ => { updateCount("conn factory"); conn.Open(); })).Bind(func));

            var connect = Connect<string>()
                .Par(openConnection.Par(beginTransaction))
                .Compose(() => mockConnectionFactory.Object);

            Expect(connect().Match(r => r, e => e.Message), EqualTo("result"));

            mockConnection.Verify(c => c.Dispose());
            mockConnection.Verify(c => c.BeginTransaction());
            mockConnection.Verify(c => c.CreateCommand());
            mockTransaction.Verify(c => c.Dispose());
            mockTransaction.Verify(t => t.Connection);
            mockCommand.VerifySet(c => c.Transaction = mockTransaction.Object);
            mockCommand.Verify(c => c.Dispose());

            Expect(executionOrder["conn factory"], EqualTo(0));
            Expect(executionOrder["trans"], EqualTo(1));
            Expect(executionOrder["command"], EqualTo(2));
            Expect(executionOrder["with cmd"], EqualTo(3));
            Expect(executionOrder["comDispose"], EqualTo(4));
            Expect(executionOrder["transDispose"], EqualTo(5));
            Expect(executionOrder["conDispose"], EqualTo(6));
        }
    }
}